x/ccv/provider/keeper/partial_set_security.go

```diff
@@ -13,16 +13,31 @@ import (
 	"github.com/cosmos/interchain-security/v5/x/ccv/provider/types"
 )
 
-// HandleOptIn prepares validator `providerAddr` to opt in to `chainID` with an optional `consumerKey` consumer public key.
+// HandleOptIn prepares validator `providerAddr` to opt in to `consumerId` with an optional `consumerKey` consumer public key.
 // Note that the validator only opts in at the end of an epoch.
-func (k Keeper) HandleOptIn(ctx sdk.Context, chainID string, providerAddr types.ProviderConsAddress, consumerKey string) error {
-	if !k.IsConsumerProposedOrRegistered(ctx, chainID) {
+func (k Keeper) HandleOptIn(ctx sdk.Context, consumerId string, providerAddr types.ProviderConsAddress, consumerKey string) error {
+	phase, found := k.GetConsumerIdToPhase(ctx, consumerId)
+	if !found || phase == Stopped {
 		return errorsmod.Wrapf(
-			types.ErrUnknownConsumerChainId,
-			"opting in to an unknown consumer chain, with id: %s", chainID)
+			types.ErrInvalidPhase,
+			"opting in to an unknown (or stopped) consumer chain, with id: %s", consumerId)
 	}
 
-	k.SetOptedIn(ctx, chainID, providerAddr)
+	registrationRecord, found := k.GetConsumerIdToRegistrationRecord(ctx, consumerId)
+	if !found {
+		return errorsmod.Wrapf(
+			types.ErrUnknownConsumerId,
+			"opting in to an unknown consumer chain, with id: %s", consumerId)
+	}
+
+	optedInToConsumerId, found := k.IsValidatorOptedInToChain(ctx, providerAddr, registrationRecord.ChainId)
+	if found && consumerId != optedInToConsumerId {
```

Comment:
This means that we allow validators to opt in twice on the same consumerId?

Reply:
True, this was the case.
I removed the `consumerId != optedInConsumerId` so they cannot opt in to the same chain again.